presto-orc/src/main/java/com/facebook/presto/orc/writer/MapFlatColumnWriter.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.writer;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.ColumnarMap;
+import com.facebook.presto.orc.ColumnWriterOptions;
+import com.facebook.presto.orc.DwrfDataEncryptor;
+import com.facebook.presto.orc.metadata.ColumnEncoding;
+import com.facebook.presto.orc.metadata.CompressedMetadataWriter;
+import com.facebook.presto.orc.metadata.MetadataWriter;
+import com.facebook.presto.orc.metadata.statistics.ColumnStatistics;
+import com.facebook.presto.orc.stream.StreamDataOutput;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.common.block.ColumnarMap.toColumnarMap;
+import static com.facebook.presto.orc.metadata.ColumnEncoding.ColumnEncodingKind.DWRF_MAP_FLAT;
+import static com.facebook.presto.orc.metadata.CompressionKind.NONE;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MapFlatColumnWriter
+        implements ColumnWriter
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(MapFlatColumnWriter.class).instanceSize();
+    private final int column;
+    private final boolean compressed;
+    private final ColumnEncoding columnEncoding;
+    private final CompressedMetadataWriter metadataWriter;
+
+    private boolean closed;
+
+    public MapFlatColumnWriter(int column, ColumnWriterOptions columnWriterOptions, Optional<DwrfDataEncryptor> dwrfEncryptor, MetadataWriter metadataWriter)
+    {
+        // TODO: Flat map needs to get factories for key and value writers, instead of already initialized writers
+        checkArgument(column >= 0, "column is negative");
+        requireNonNull(columnWriterOptions, "columnWriterOptions is null");
+        this.column = column;
+        this.compressed = columnWriterOptions.getCompressionKind() != NONE;
+        this.columnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0); // TODO Use c'tor with additionalSequenceEncodings
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+    }
+
+    @Override
+    public List<ColumnWriter> getNestedColumnWriters()
+    {
+        // TODO: implement me
+        return ImmutableList.of();
+    }
+
+    @Override
+    public Map<Integer, ColumnEncoding> getColumnEncodings()
```

Comment:
FlatMap has encodings per nodeid, sequence , so this needs to change in the future. Something like a Map<NodeSequence, Encoding> would be nice!

Reply:
The ColumnEncoding, created in the constructor, has support for additionalSequenceEncodings. I'll check if it also fits the writer case.

```
    // DWRF supports the concept of sequences.
    // A column can be modeled as multiple sequences that are independently encoded.
    // For example, for a flat map column, each key will have a
    // separate value stream with its own column encoding.
    // These additional sequence IDs start from 1 and may not be consecutive, for example, the file may be updated to
    // remove keys from the flat map without changing the sequence IDs associated with each key.
    // Sorted so that when we iterate over the map the DwrfSequenceEncodings are returned in ascending Sequence ID order
    private final Optional<SortedMap<Integer, DwrfSequenceEncoding>> additionalSequenceEncodings;
```